server.js

```diff
@@ -5,6 +5,101 @@ const app = express();
 
 app.use(cors());
 
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+//Create a new message
+app.post("/messages", (req, res) => {
+  let newPost = {
+    id: messages.length,
+    from: req.body.from,
+    text: req.body.text,
+    timeSent: dateTime,
+  };
+  if (newPost.from !== "" && newPost.text !== "") {
+    messages.push(newPost);
+    res.json(messages);
+  } else {
+    res.status(400).json(`Please fill all the fields`);
+  }
+});
+
+// Delete a message, by ID
+app.delete("/messages/delete/:id", (req, res) => {
+  let id = Number(req.params.id);
+  let index = messages.findIndex((message) => message.id === id);
+  if (index) {
+    messages.splice(index, 1);
+    res.json(`message with id: ${id} has been deleted`);
+  } else {
+    res.status(404).json("Please enter a valid ID to delete");
+  }
+});
+//Read all messages
+app.get("/messages", (req, res) => {
+  res.json(messages);
+});
+
+// add support for the client to be able to update a message's text or from property.
+app.put("/messages/:id", (req, res) => {
+  let id = Number(req.params.id);
+  let newText = req.body.text;
+  let found = messages.find((message) => message.id == id);
+  let updatedMsg = [];
+  messages.forEach((message) => {
+    if (message.id === found.id) {
+      message.text = newText;
+      updatedMsg.push(message);
+    } else {
+      updatedMsg.push(message);
+    }
+  });
+  res.json({ updatedMsg });
+});
+
+//Read only messages whose text contains a given substring
+app.get("/messages/search", (req, res) => {
+  let text = req.query.text;
+  if (text !== undefined) {
+    let result = messages.filter((message) =>
+      message.text.toLowerCase().includes(text.toLowerCase())
+    );
+
+    if (result.length < 1) {
+      res.status(400).json(`Please enter a valid search term`);
+    } else {
+      res.json(result);
+    }
+  }
+});
+
+//Read only the most recent 10 messages
+app.get("/messages/latest", (req, res) => {
+  let latestMessages = messages.slice(-10);
```

Comment:
So because the messages are always in chronological order as they are being stored after each other in the order of arrival therefore we can relly on the natural order of the messages.
However, let's assume it wouldn't be the case. What we can do is sorting the messages by timeSent for first and then slice them:
`messages.sort((a, b) => a.timeSent - b.timeSent).slice(-10)`
https://stackoverflow.com/questions/16690035/sort-date-in-javascript

Reply:
Just realised suddenly that you store the date in text format, so my suggested sorting method is wrong.